libraries/abstractions/platform/freertos/iot_network_udp.c

```diff
@@ -0,0 +1,534 @@
+    {
+        IotLogWarn( " Async receive indicated failure\n" );
+    }
+    else
+    {
+        while( bytesRemaining > 0 )
+        {
+            socketStatus = SOCKETS_RecvFrom( pNetworkConnection->socket,
+                                             pBuffer + bytesReceived,
+                                             bytesRemaining,
+                                             0,
+                                             &serverAddress,
+                                             &addrLength );
+
+            if( socketStatus <= 0 )
+            {
+                IotLogError( "Failed to receive requested bytes: %d, received: %d socketStatus: %d", bytesRequested, bytesReceived,
+                             socketStatus );
+                break;
+            }
+
+            bytesRemaining -= socketStatus;
+            bytesReceived += socketStatus;
+            configASSERT( bytesReceived + bytesRemaining == bytesRequested );
+        }
+    }
+
+    if( ( bytesReceived > 0 ) && ( pNetworkConnection->strictServerAddrCheck ) )
+    {
+        /* if strict server address check is enabled, only accept packets from the server that was used for Send */
+        if( serverAddress.ulAddress != pNetworkConnection->serverAddr.ulAddress )
+        {
+            IotLogError( "Server address does not match and strict server address check enabled." );
+            bytesReceived = 0;
+        }
+    }
+    /* if Receive was called as a result of asynch callback, set bufferbytes valid to false.*/
+    if ((xTaskGetCurrentTaskHandle() == pNetworkConnection->receiveTask) )
+    {
+        pNetworkConnection->bufferedByteValid = false;
+    }
+    
+
+    return bytesReceived;
+}
+/*-----------------------------------------------------------*/
+
+IotNetworkError_t IotNetworkUdp_Close( void * pConnection )
+{
+    int32_t socketStatus = SOCKETS_ERROR_NONE;
+
+    /* Cast network connection to the correct type. */
+    _networkConnection_t * pNetworkConnection = ( _networkConnection_t * ) pConnection;
+
+    if( pNetworkConnection->socket != SOCKETS_INVALID_SOCKET )
+    {
+        IotLogWarn( " Close failed: Invalid Socket " );
+        socketStatus = SOCKETS_SOCKET_ERROR;
+    }
+    else
+    {
+        /* Call Secure Sockets shutdown function to close connection. */
+        socketStatus = SOCKETS_Shutdown( pNetworkConnection->socket,
+                                         SOCKETS_SHUT_RDWR );
+
+        if( socketStatus != SOCKETS_ERROR_NONE )
+        {
+            IotLogWarn( "Failed to close connection." );
+        }
+
+        /* If receive task was started , indicate SHUT_DOWN */
+        if( pNetworkConnection->receiveCallback != NULL )
+        {
+            /* Set the shutdown flag. */
+            ( void ) xEventGroupSetBits( ( EventGroupHandle_t ) &( pNetworkConnection->connectionFlags ),
+                                         _FLAG_SHUTDOWN );
```

Comment:
the order here is different than order in ```IoTNetworkUdp_Destroy```. In that function, the task is signaled and we wait for the task to exit, before we destroy the connection. Here we shutdown the socket, and then signal the task to shutdown. 

Why the difference?

Reply:
Here the goal is to shutdown the socket in both direction and wait for recv task to exit .  Destroy call destroys the connection, hence receive task must exit (if it is running). UDP can only have destroy and does not need close(), but since API provides the call, it is intended  for 'two step' close. (A destroy call is still needed)

Reply:
"Here the goal is to shutdown the socket in both direction and wait for recv task to exit ." Thats exactly my point. I dont see the logic to wait for the task to exit, it seems we only set the bit and return from this function, without waiting to confirm that the task exited.

Reply:
IotNetworkUdp_Close() has to be followed by IoTNetworkUdp_Destroy(). _FLAG_SHUTDOWN will break the while loop in the recv task.  oTNetworkUdp_Destroy() will then make sure task has exited.

Reply:
After calling IotNetworkUdp_Close() the receive task may not exit, to ensure receive task is exited we call IotNetworkUdp_Destroy() to prevent any further connection.